types/address/hash.go

```diff
@@ -59,10 +61,15 @@ func Compose(typ string, subAddresses []Addressable) ([]byte, error) {
 }
 
 // Module is a specialized version of a composed address for modules. Each module account
-// is constructed from a module name and module account key.
-func Module(moduleName string, key []byte) []byte {
+// is constructed from a module name and a sequence of derivation keys (at least one
+// derivation key must be provided).
+func Module(moduleName string, derivationKey []byte, derivationKeys ...[]byte) []byte {
```

Comment:
I was thinking to "merge" derivationKey into derivationKeys, but then we would need to panic if someone will provide an empty list.

Reply:
I think as is is great, requires at least 1 derivationKey implicitly which is communicated to the caller.

Reply:
I don't see any reason to require at least one derivation key for modules. This is inconsistent with what we've discussed for ADR 033. There is always the root module address with no derivation keys.

```suggestion
func Module(moduleName string, derivationKeys ...[]byte) []byte {
```

Reply:
No. That uses 20 byte addresses

Reply:
And if it were the same implementation that would actually be a good thing because it's the same intention

Reply:
so, if `derivationKeys` we fallback to `authtypes.NewModuleAddress`? That sounds good.

PS: Ideally I would prefer to keep all modules in the `module` namespace, but I think we need to stay with `authtypes.NewModuleAddress`.

Reply:
BTW, with conflict I meant that users would have 2 different functions with 2 different results. But if we handle the special case as described above then we are OK.

Reply:
No I think this function should still use 32 byte addresses based on what the input from the cryptographer was.

Reply:
My preference is to keep backwards compatibility of existing module accounts.

```go
func Module(moduleName string, derivationKeys ...[]byte) []byte {
  if len(derivationKeys) == 0 { return authtypes.NewModuleAddress() }
  else { /* do like current code */ }
}
```

so it's:
- 20 byte address for root module accounts
- 32 byte addresses for sub/derived module accounts

Reply:
So we need to make decision.

Reply:
up

Reply:
Up again, I'm going to update this PR by implementing https://github.com/cosmos/cosmos-sdk/pull/14017#discussion_r1040287293  (which is what @amaurym seconded: https://github.com/cosmos/cosmos-sdk/pull/14017#discussion_r1041092641)

Reply:
SGTM